index.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    arr = to_a
+    arr.to_enum unless block_given?
+    n = 0
+    while arr.length > n
+      yield(arr[n])
+      n += 1
+    end
+  end
+
+  def my_each_with_index
+    arr = to_a
+    return arr.to_enum unless block_given?
+
+    n = 0
+    while arr.length > n
+      yield(arr[n], n)
+      n += 1
+    end
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    res = []
+    my_each do |v|
+      res << v if yield(v) == true
+    end
+    res
+  end
+
+  def my_all?
+    check = true
+    if block_given?
+      my_each { |i| check = false unless yield(i) }
+    else
+      my_all? { |obj| obj }
+    end
+    check
+  end
+
+  def my_any?
+    check = false
+    if block_given?
+      my_each { |i| check = true if yield(i) }
+    else
+      my_any? { |obj| obj }
+    end
+    check
+  end
+
+  def my_none?
+    check = true
+    if block_given?
+      my_each { |i| check = false if yield(i) }
+    else
+      my_none { |obj| obj }
+    end
+    check
+  end
+
+  def my_count(*arg)
+    count = 0
+    unless block_given?
+      if arg.empty?
+        my_count { |i| i }
+      else
+        my_count { |i| arg[0] == i }
+      end
+    end
+    my_each { |i| count += 1 if yield(i) }
+    count
+  end
+
+  def my_map(&proc)
+    return arr.to_enum unless block_given?
+
+    result = []
+    my_each { |i| result << proc.call(i) }
+    result
+  end
+
+  def my_inject(init = nil)
+    n = 0
+    arr = to_a
+    acc = init ? init + arr[0] : arr[0]
```

Comment:
- [ ] my_inject combines all elements of enum by applying a binary operation, specified by a block

```
irb(main):104:0> [1,2,3].my_inject(2) {|acc, n| acc * n}
=> 18
irb(main):105:0> [1,2,3].inject(2) {|acc, n| acc * n}
=> 12
irb(main):106:0> [1,2,3].my_inject(2) {|acc, n| acc + n}
=> 8
irb(main):107:0> [1,2,3].inject(2) {|acc, n| acc + n}
=> 8
irb(main):110:0> [1,2,3].inject(10) {|acc, n| acc * n}
=> 60
irb(main):111:0> [1,2,3].my_inject(10) {|acc, n| acc * n}
=> 66

```

- [ ] my_inject when a symbol is specified combines each element of the collection by applying the symbol as a named method

```
irb(main):099:0> [1,2,3].inject(:+)
=> 6
irb(main):100:0> [1,2,3].my_inject(:+)
Traceback (most recent call last):
....
NoMethodError (undefined method `+' for :+:Symbol)


```